src/k2/tso/service/TSOController.cpp

```diff
@@ -219,6 +219,20 @@ void TSOService::TSOController::RegisterGetTSOWorkersURLs()
 
 void TSOService::TSOController::HeartBeat()
 {
+    // delayed heartbeat detection and logging
+    uint64_t nowCnt = now_nsec_count();
+
+    if ((nowCnt - _lastHeartBeat) > (uint64_t)( 2 * (nsec(_heartBeatTimerInterval()).count())))
+    {
+        if (_lastHeartBeat != 0)
+        {
+            K2LOG_I(log::tsoserver, "HeartBeat delayed more than two preset intervals, _lastHeartBeat:{}, preset duration in nano sec: {}", 
+                _lastHeartBeat,  nsec(_heartBeatTimerInterval()).count());

```

Comment:
you can just log _heartBeatTImerInterval(). I added conversion formatters for all timestamps as well so these should log nicely.

Reply:
.

Reply:
discussed - nanosec precision is needed so the change's good